File: csharp/SchoolData/Controllers/AssignController.cs
using Gb.Homework.SchoolData.Model;
using Gb.Homework.SchoolData.Service;
using Gb.Homework.SchoolData.View;

namespace Gb.Homework.SchoolData.Controllers
{

    public class AssignController
    {
        private readonly ISchoolUserService<Teacher> teacherService;
        private readonly ISchoolUserService<Student> studentService;
        private readonly AssignView view;

        public AssignController(ISchoolUserService<Teacher> teacherService, ISchoolUserService<Student> studentService, AssignView assignView)
        {
            this.teacherService = teacherService;
            this.studentService = studentService;
            view = assignView;
        }

        public void Run()
        {
            view.ShowTitle();

            foreach (var item in teacherService.GetAllUsers())
            {
                item.ClearGroup();
            }

            Stack<Student> students = new Stack<Student>();

            foreach (var item in studentService.GetAllUsers())
            {

                students.Push(item);
            }


            while (students.Count > 0)
            {
                foreach (var t in teacherService.GetAllUsers())
                {
                    if (students.Count == 0)
                        break;

                    t.addStudent(students.Pop());
                }
            }


            foreach (var t in teacherService.GetAllUsers())
            {
                view.ShowTeacher(t);
            }
        }
    }
}

File: SchoolData/src/main/java/org/gb/homework/schooldata/controller/AssignController.java
package org.gb.homework.schooldata.controller;

import org.gb.homework.schooldata.model.Student;
import org.gb.homework.schooldata.model.Teacher;
import org.gb.homework.schooldata.service.UserService;
import org.gb.homework.schooldata.view.AssignView;

import java.util.Stack;

public class AssignController {
    private final UserService<Teacher> teacherService;
    private final UserService<Student> studentService;
    private final AssignView view;

    public AssignController(UserService<Teacher> teacherService, UserService<Student> studentService, AssignView assignView) {
        this.teacherService = teacherService;
        this.studentService = studentService;
        this.view = assignView;
    }

    public void run() {
        view.showTitle();

        for (var t : teacherService.getAllUsers()) {
            t.clearGroup();
        }

        Stack<Student> students = new Stack<>();
        students.addAll(studentService.getAllUsers());

        while (!students.isEmpty()) {
            for (var t : teacherService.getAllUsers()) {
                if (students.isEmpty())
                    break;

                t.addStudent(students.pop());
            }
        }


        for (var t : teacherService.getAllUsers()) {
            view.showTeacher(t);
        }
    }
}
